### ISSUE TITLE
Converter functions remove type annotations from __init__

### ISSUE THREAD

**Issue by nosewings** [python-attrs/attrs#694]
Example:

```python
from typing import Union
import attr

def to_int(x: Union[str, int]) -> int:
    return int(x)

@attr.s
class Example:
    x: int = attr.ib(converter=to_int)
```

My expectation is that the generated `__init__` should take a parameter `x: Union[str, int]`. Instead, the generated `__init__` simply takes an untyped parameter `x`.

It seems like this could be solved by looking at the type annotations attached to the converter method and then annotating `__init__` appropriately.

If this is a desirable feature, I can submit a PR.

**Comment by hynek** [comment]
I feel like this has come up before, but just to be clear: are you talking about whatever is in `__annotations__` or about the behavior of mypy in these cases?

**Comment by nosewings** [comment]
I'm just talking about `__annotations__`. In the previous example, we currently have
```python
>>> Example.__init__.__annotations__
{'return': None}
```
whereas I would expect
```python
>>> Example.__init__.__annotations__
{'x': typing.Union[str, int], 'return': None}
```

**Comment by euresti** [comment]
Ah. This happens here

https://github.com/python-attrs/attrs/blob/48755905ee8088da927a6dc583d4962f7ca16b9b/src/attr/_make.py#L2115

```
        if a.init is True and a.converter is None and a.type is not None:
            annotations[arg_name] = a.type
```

**Comment by hynek** [comment]
So the q is why did the `a.converter is None` check instead of copying it over? It should be noted, that the fix needs to special-case the use of the `pipe()` converter.

**Comment by hynek** [comment]
@nosewings are you still game for a PR?

**Comment by nosewings** [comment]
Yes. School has kept me busy, but I think I can get the PR out within the next few days.

**Comment by nosewings** [comment]
I will note that a choice has to be made about what to do if both a converter and an explicit type annotation or `type` argument are present. From an implementation perspective, the simplest thing to do is to let a `type` argument override any converter annotations. However, that might have some unintuitive consequences. Consider the following:

```python
import attr

def int2str(x: int) -> str:
    return str(x)

@attr.s
class A:
    a: str = attr.ib(converter=int2str)
```

A plausible reading of this code's intent is that the _attribute_ `a` should have type `str`, but the _argument_ `a` should have type `int`. However, given the implementation I've suggested above, the `a: str` annotation would override the converter's annotation, so `__init__` would take an argument `a: str`.

**Comment by hynek** [comment]
I suspect this muddying might be reason, why we have punted on it?

I mean the correct annotation for `a` in init would be the argument type of the _first_ converter (in this case `int`), correct?

**Comment by nosewings** [comment]
@hynek Different users could have different intents. In any case, I feel like it would be surprising if the `type` argument could be overridden implicitly

In any case, I've [submitted a PR](https://github.com/python-attrs/attrs/pull/710). Of course, it's completely reasonable if the issue I've raised above makes you wary of implementing this change.

**Comment by sscherfke** [comment]
IMHO, the type of the `value` argument of *all* of an attribute converters and validators must be the same. 
This type should/could then be used to annotate the type of that attribute in `__init__()`.

Vise versa, the return type of all converers must be the same and must also match an explicitly annotated type for an attribute. This is the type that a "getattr()" would return.

If the converters and validators do use type annotations and if these annotations do not mach, attrs should raise a TypeError while creating the class. This would avoid the need for attrs to handle ambiguities in the user's code. :)

Here is an example in pseudo code for illustration:
```python
T_IN = Type(...)
T_OUT = Type(...)

def validator(value: T_IN) -> None:
    pass

def converter(value: T_IN) -> T_OUT:
    return T_OUT(value)

@attr.define
class C:
    x: T_OUT = attr.field(converter=converter, validator=validator)

# Resulting/generated class:
class C:
    def __init__(self, x: T_IN): -> None:
        self.__validate_x(x)
        self.x: T_OUT = self.__convert_x(x)
```

**Comment by hynek** [comment]
> If the converters and validators do use type annotations and if these annotations do not mach, attrs should raise a TypeError while creating the class. This would avoid the need for attrs to handle ambiguities in the user's code. :)

Oof, I think this leads down a dangerous path with types that are actually compatible even though the look differently. This should be left to the type checking crowd (i.e. mypy) IMHO.

**Comment by hynek** [comment]
This has been fixed by #710
